src/contact.rs

```diff
@@ -0,0 +1,1120 @@
+                ).ok();
+                }
+                sth_modified = Modifier::Modified;
+            }
+        } else {
+            if sql::execute(
+                context,
+                &context.sql,
+                "INSERT INTO contacts (name, addr, origin) VALUES(?, ?, ?);",
+                params![name.as_ref(), addr, origin,],
+            )
+            .is_ok()
+            {
+                row_id = sql::get_rowid(context, &context.sql, "contacts", "addr", addr);
+                sth_modified = Modifier::Created;
+            } else {
+                error!(context, 0, "Cannot add contact.");
+            }
+        }
+
+        Ok((row_id, sth_modified))
+    }
+
+    /// Add a number of contacts.
+    ///
+    /// Typically used to add the whole address book from the OS. As names here are typically not
+    /// well formatted, we call `normalize()` for each name given.
+    ///
+    /// No email-address is added twice.
+    /// Trying to add email-addresses that are already in the contact list,
+    /// results in updating the name unless the name was changed manually by the user.
+    /// If any email-address or any name is really updated,
+    /// the event `DC_EVENT_CONTACTS_CHANGED` is sent.
+    ///
+    /// To add a single contact entered by the user, you should prefer `Contact::create`,
+    /// however, for adding a bunch of addresses, this function is _much_ faster.
+    ///
+    /// The `adr_book` is a multiline string in the format `Name one\nAddress one\nName two\nAddress two`.
+    ///
+    /// Returns the number of modified contacts.
+    pub fn add_address_book(context: &Context, adr_book: impl AsRef<str>) -> Result<usize> {
+        let mut modify_cnt = 0;
+
+        for chunk in &adr_book.as_ref().lines().chunks(2) {
+            let chunk = chunk.collect::<Vec<_>>();
+            if chunk.len() < 2 {
+                break;
+            }
+            let name = chunk[0];
+            let addr = chunk[1];
+            let name = normalize_name(name);
+            let (_, modified) = Contact::add_or_lookup(context, name, addr, Origin::AdressBook)?;
+            if modified != Modifier::None {
+                modify_cnt += 1
+            }
+        }
+        if modify_cnt > 0 {
+            context.call_cb(Event::CONTACTS_CHANGED, 0 as uintptr_t, 0 as uintptr_t);
+        }
+
+        Ok(modify_cnt)
+    }
+
+    /// Returns known and unblocked contacts.
+    ///
+    /// To get information about a single contact, see dc_get_contact().
+    ///
+    /// `listflags` is a combination of flags:
+    /// - if the flag DC_GCL_ADD_SELF is set, SELF is added to the list unless filtered by other parameters
+    /// - if the flag DC_GCL_VERIFIED_ONLY is set, only verified contacts are returned.
+    ///   if DC_GCL_VERIFIED_ONLY is not set, verified and unverified contacts are returned.
+    /// `query` is a string to filter the list.
+    pub fn get_all(
+        context: &Context,
+        listflags: u32,
```

Comment:
Is this flag use "Rusty"?  I've never looked into how flags would look the rusty way.

Reply:
https://docs.rs/bitflags/1.1.0/bitflags/ is how I do nice flags in rust, but haven't done that at all in the code yet.